### ISSUE TITLE
disallowing `Tuple{Union{}}` makes `hasmethod` less powerful.

### ISSUE THREAD

**Issue by oscardssmith** [JuliaLang/julia#56279]
In 1.9: it was possible to qurery the method table to see whether a given function has a 2 argument method: e.g. 
```
julia> g(u::Float64,p) = u

julia> hasmethod(g, Tuple{Int, Int})
false

julia> hasmethod(g, Tuple{Union{}, Union{}})
true
```
However on 1.10, we can't express ` Tuple{Union{}, Union{}}` so we can't express it like this (instead you have to do something like searching through all `methods` which can't be done at compile-time.

This is closely related to https://github.com/JuliaLang/julia/issues/52385, but IMO is sufficiently different to be worth separating into a separate issue.

**Comment by vtjnash** [comment]
```
  julia> hasmethod(g, Tuple{Union{}, Union{}})
  true
  ```

This result seems like bug: it probably should have returned false, since Julia may assume that the method list queries obeys subtyping transitivity, while that method cannot be called / does not exist for all supertypes of the input query. That is a bug in v1.9 though, and gets fixed (not a regression) in v1.10.

**Comment by topolarity** [comment]
> that method cannot be called / does not exist for all supertypes of the input query

What supertype can it not be called for?

**Comment by topolarity** [comment]
But that's also true for:
```julia
julia> g(::Union{Int,Float64}) = 13
julia> hasmethod(g, Tuple{Int})
true
```

`Tuple{Int,Float32}` is a supertype of `Tuple{Int}` and isn't associated with any method of `g`

I thought that `hasmethod` is supposed to return `true` iff the argtype provided is completely covered by any method definition.
(IIUC, that's what @oscardssmith is expecting here too.)

**Comment by topolarity** [comment]
Is the disagreement about whether `Tuple{Float64, Any}` is a supertype of `Tuple{Union{},Union{}}`?

**Comment by vtjnash** [comment]
Ah right. Maybe the issue is that it may return false because it determines is ambiguous which method to (not) call, so it would be wrong to ever use as an approximation of whether a method exists with that number of arguments?

**Comment by vtjnash** [comment]
> true iff the argtype provided is completely covered by any method definition

Covered by exactly one method is the actual definition of this query, since that is the definition used by invoke, and this does the same query as invoke during runtime (during inference, it does the same query also as invoke, including first discarding the possibility of any parameter being Union{})

**Comment by oscardssmith** [comment]
One resolution for this would be a new `hasmethod_n_args(n)` intrinsic. For anyone interested, the reason I want this capability is to be able to write https://github.com/SciML/SciMLBase.jl/blob/3c1211ae5cc030dbd0d991f0ce26fe7521b904d1/src/utils.jl#L246 in a better (preferably compile time-executable) way.

**Comment by topolarity** [comment]
> Covered by exactly one method is the actual definition of this query, since that is the definition used by invoke, and this does the same query as invoke during runtime

Right, I see - this is emulating `invoke` so for inhabited types `hasmethod` clearly must be true if there is a most-specific, covering method

I don't think inhabited-ness is doing any work in that definition though TBH. The query over uninhabited types still seems reasonable to me (is there a most-specific, covering method) - I don't see where we discard any uninhabited parameter types?

Of course, that's assuming that https://github.com/JuliaLang/julia/pull/54792 lands one day and not all uninhabited types are type-equal to `Union{}` because otherwise this query is especially useless (it just checks whether the function has 1 method)

> Ah right. Maybe the issue is that it may return false because it determines is ambiguous which method to (not) call, so it would be wrong to ever use as an approximation of whether a method exists with that number of arguments?

That's true - it seems like you'd want to throw out specificity and just check for covering methods for the query @oscardssmith wants

**Comment by oscardssmith** [comment]
With the following example, I do think a `hasmethod` type version of this won't work well.
```
julia> g(u::Float64,p) = u
g (generic function with 1 method)

julia> g(u::Int,p) = u
g (generic function with 2 methods)

julia> hasmethod(g, Tuple{Union{}, Union{}})
false
```

**Comment by bvdmitri** [comment]
> One resolution for this would be a new hasmethod_n_args(n) intrinsic. 

This approach would be much clearer, as `hasmethod(g, Tuple{Union{}, Union{}})` is quite obscure for someone unfamiliar with the underlying mechanics (like me). Just by looking at it, it's not intuitive that this checks for a method with two arguments and code just smells. It feels "hacky" at best. IMO most users would find this confusing (could be wrong though), since you need to understand the semantics around `Union{}`, its relationship to other types, and its interaction with `Tuple{}`. A cleaner alternative could be indeed a separate function or even maybe smth like `hasmethod_n_args(g, Val(2))` if you need this check at compile time.

**Comment by JeffBezanson** [comment]
Should this issue be renamed to a request for `hasmethod_n`?

**Comment by frankwswang** [comment]
I want to point out that it might be desirable to use `Tuple{Vararg{Union{}}}` as the second argument of `hasmethod` to only detect the existence of certain keyword arguments. For example:
```
julia> f(a::Int, b::Int; c::Int=1) = (a, b, c)
f (generic function with 1 method)

julia> hasmethod(f, Tuple{Union{}, Union{}}, (:c,)) # At Julia 1.9.4
true
```
So, I'm not sure if this scenario can be resolved in 1.10+. Although adding another method can already break the result:
```
julia> f(a::Int, b::Float64) = (a, b)
f (generic function with 2 methods)

julia> hasmethod(f, Tuple{Union{}, Union{}}, (:c,))
false
```

**Comment by vtjnash** [comment]
Yes, as you observe about it in v1.9, it returned an essentially random Boolean, since you were asking a query that is likely impossible to answer correctly

**Comment by oscardssmith** [comment]
@vtjnash and @JeffBezanson. would you two be OK with `hasmethod_n`? We once again ran up against the lack of it.
